src/stepfunctions/steps/states.py

```diff
@@ -541,7 +541,7 @@ def __init__(self, state_id, **kwargs):
         Args:
             state_id (str): State name whose length **must be** less than or equal to 128 unicode characters. State names **must be** unique within the scope of the whole state machine.
             iterator (State or Chain): State or chain to execute for each of the items in `items_path`.
-            items_path (str, optional): Path in the input for items to iterate over. (default: '$')
+            items_path (str or Placeholder, optional): Path in the input for items to iterate over. (default: '$')
```

Comment:
Needs a unit test. Do the existing fields: `input_path`, `result_path`, and `output_path` even support Placeholders today? ItemsPath is similar to those in that it's key is never suffixed by `.$` like Parameter keys referencing a Path. Its value is always treated a JSONPath, never a regular string. 

In addition to `items_path`, https://github.com/aws/aws-step-functions-data-science-sdk-python/issues/101 is also trying to use `ExecutionInput` for `input_path`.

Reply:
You're right - i forgot to remove the Placeholder mention from an earlier commit. 
We don't currently support Placeholders for `items_path`, `input_path`, `result_path`, and `output_path`.
I can add support in the next commit

Reply:
I can add support in my next commit

Reply:
I tested in the AWSConsole and currently, we don't support Placeholders for `items_path`, `input_path`, `result_path` and `output_path`

Tested with a state machine based on [Map State Example](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-map-state.html)
```
{
  "StartAt": "MapExample01",
  "States": {
    "MapExample01": {
      "InputPath": "$$.Execution.Input['input_path']",
      "ResultPath": "$.detail.shipping",  # Note: State machine would fail to create with execution input
      "ItemsPath": "$$.Execution.Input['items_path']",
      "OutputPath": "$$.Execution.Input['output_path']",
      "Type": "Map",
      "End": true,
      "Iterator": {
        "StartAt": "Validate",
        "States": {
          "Validate": {
            "Resource": <resource_arn>,
            "Type": "Task",
            "End": true
          }
        }
      }
    }
  }
}
```

With execution input:

```
{
  "ship-date": "2016-03-14T01:59:00Z",
  "detail": {
    "delivery-partner": "UQS",
    "shipped": [
      {
        "prod": "R31",
        "dest-code": 9511,
        "quantity": 1344
      },
      {
        "prod": "S39",
        "dest-code": 9511,
        "quantity": 40
      }
    ]
  },
  "input_path": "$.detail",
  "items_path": "$.shipped",
  "output_path": "$.detail.shipped"
}
```
Execution would fail as paths could not be applied to execution input

I will remove the Placeholder mention from the doc

Reply:
That's not true. InputPath, ItemsPath, OutputPath, and ResultPath must be JSONPath values. The exception you got during execution is because the payload between input/output processing steps didn't have the right data.

```
Raw input -> InputPath  -> ItemsPath  ->  Parameters  -> ResultPath  -> Output Path
```

The example you used doesn't work because the value eferenced in the context object is a JSONPath string instead of an array. For ItemsPath, the dereferenced runtime value is the literal string "$.shipped". Which is not an array. Step Functions doesn't treat the value as another JSONPath to interpolate again. You need to set `"InputPath": "$.shipped"` and `"ItemsPath": "$.items"` in the state machine definition or reference them from the context objext ($$.Execution.Input.shipped).

"Placeholder" is a data science SDK concept that doesn't exist in ASL that makes it easier to reference dynamic variables in a pythonic way without writing JSONPath.

Reply:
Thanks for your input!
I confirm that it does work to use the reference directly

```
{
  "StartAt": "MapExample01",
  "States": {
    "MapExample01": {
      "InputPath": "$$.Execution.Input['input_path']",
      "ResultPath": "$.detail.shipped",
      "ItemsPath": "$.shipped",
      "OutputPath": "$.detail.shipped",
      "Type": "Map",
      "End": true,
      "Iterator": {
        "StartAt": "Validate",
        "States": {
          "Validate": {
            "Resource": <resource_arn>,
            "Type": "Task",
            "End": true
          }
        }
      }
    }
  }
}
```

with this execution input

```
{
  "ship-date": "2016-03-14T01:59:00Z",
  "detail": {
    "delivery-partner": "UQS"
  },
  "input_path": {
    "delivery-partner": "UQS",
    "shipped": [
      {
        "prod": "R31",
        "dest-code": 9511,
        "quantity": 1344
      },
      {
        "prod": "S39",
        "dest-code": 9511,
        "quantity": 40
      }
    ]
  }
}
```

Reply:
Change to support `itemspath` and `inputpath` with placeholders will be added with the next commit